### PR TITLE
S24/sayi/add user management endpoints

### PR DESCRIPTION
**Pull request by sthuray**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Task-Board-db95cd7b93f245f78ee85e3a8a6a316d)

## Implementation description
* Changed UpdateUserDTO/UpdateUserDtoValidator so that users can update a user without providing every field
* If a behaviourist tries to update a user and they update a value that they don't have access to, entire request is invalidated

## Steps to test
GET
1. No queries (get all users)
UME - GET - all users

2. Query with both userId & email
UME - GET - email and userId query

3. Invalid userId
UME - GET - invalid userId

4. UserId not in database
UME - GET - nonexisting userid

5. Get with userId (successful)
UME - GET by userid

6. Invalid email
UME - GET - invalid email

7. Email not in database
UME - DELETE - nonexisting email

8. Get with email (successful)
UME - GET by email

PUT:
1. (Behaviourist) userId not in database (similarly, test invalid userId)
UME - AM - userId not found

2. (Behaviourist) Update only invalid fields
UME - AM - Update only invalid fields

3. (Behaviourist) Update invalid & valid fields
UME - AM - Update a few valid:invalid fields

4. (Behaviourist) Update valid fields
UME - AM - Successful Update

5. (Admin) Update all fields
UME - Admin - Success

DELETE:
1. Query with both userId & email
UME - DELETE - both params

2. Query without parameters
UME - DELETE - no params

3. Invalid userId
UME - DELETE - Invalid userid

4. userId not in database
UME - DELETE - nonexisting userid

5. Delete user with "Active" status by userId/email
UME - DELETE - active by userid

6. Delete Animal Behaviourist with "Inactive" status by userId
UME - DELETE - inactive behaviourist by email

6. Invalid email
UME - DELETE - invalid email

7. Email not in database
UME - DELETE - nonexisting email

8. Delete volunteer with "Invited" status (by email)
UME - DELETE - invited volunteer by id

9. Delete as a Behaviourist
UME - DELETE - delete as behaviourist

## What should reviewers focus on?
* Refactored try catch statements in PUT 
* Set filtering for PUT
* Querying with a non-existent email doesn't return a NotFoundError (add this to emailService?)
* Querying with userId=10. gets through since Number(10.) returns 10, doesn't cause any problems, only visible when returning "userId 10. not found in database" (still checks database for userId 10 here)

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by trinity-y — APPROVED** [review]
looks good to me!!

**Review by JustinScitech — APPROVED** [review]
Tested out these endpoints. The changes look good to me. Should be good to merge.

File: backend/typescript/rest/userRoutes.ts [hunk 1]
```diff
@@ -119,26 +125,65 @@ userRouter.post("/", createUserDtoValidator, async (req, res) => {
...
+
   try {
-    const userId = Number(req.params.userId);
-    if (Number.isNaN(userId)) {
-      res.status(400).json({ error: "Invalid user ID" });
+    const isBehaviourist = await authService.isAuthorizedByRole(
+      accessToken,
+      new Set([Role.ANIMAL_BEHAVIOURIST]),
+    );
+    const behaviouristUpdatableSet = new Set(["skillLevel"]);
```
Comment (jerry-cheng5) [inline backend/typescript/rest/userRoutes.ts:145]: I believe behaviourists can also edit the permissions (can see all logs, can assign users to tasks)

**Review by jerry-cheng5 — COMMENTED** [review]
Great work! Just a couple minor things. fix them and you should be good to merge!

**Review by laks0407 — APPROVED** [review]
Look good to me! Great work.

**Pull request merged**